### DISCUSSION TITLE
zh/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9442] [Docs]
# zh/

探索Ultralytics YOLOv8 的完整指南，这是一个高速、高精度的对象检测和图像分割模型。安装、预测、培训教程等。

https://docs.ultralytics.com/zh/

**Comment by emptySunce** [comment]
I want to use Augment in YOLOv8 to augment my dataset. I have customized a YAML file to store my Augment parameters. How can I save the images and labels after using Augment?

**Reply by glenn-jocher** [reply]
Hey there! 👋

To save augmented images and labels using a custom YAML file for augmentation with YOLOv8, you can specify your augment parameters in your custom YAML and pass this during your training routine. However, saving augmented images directly from the internal augmentation process is not supported out of the box.

A workaround is to replicate the augmentation logic in a separate script and manually apply it to your dataset. Here’s a basic code snippet to get you started. This example assumes you're using basic image transformations:

```python
from torchvision import transforms
from PIL import Image

# Example transform
transform = transforms.Compose([
    transforms.RandomHorizontalFlip(p=0.5),
    # Add more transforms here based on your custom YAML file
])

# Load your image
image = Image.open('path/to/your/image.jpg')

# Apply the transform
augmented_image = transform(image)

# Save the augmented image
augmented_image.save('path/to/save/augmented_image.jpg')
```

This snippet uses PyTorch's torchvision transforms for augmentation, which you'll need to adjust according to the augments you defined in your YAML file. For complex augmentations specific to YOLOv8 augments not covered by torchvision, you might need to implement them manually.

I hope this helps. Let us know if you have any other questions. Happy augmenting! 😊

**Comment by mdx123456** [comment]
您好，下面给出我的完整测试代码，利用官方源码测试语义分割，报了错误：TypeError: masks() got an unexpected keyword argument 'img_gpu' 请问一下是什么原因导致的呢，请给出一个解决方案，非常感谢！
import cv2
import torch
from ultralytics import YOLO
from ultralytics.utils.plotting import colors, Annotator
from ultralytics.data.augment import LetterBox

s_mode = 1  # 0: semantic, 1: instance
cap = cv2.VideoCapture('1.mp4')

model = YOLO('yolo11n-seg.pt')
names = model.names

w, h, fps = (int(cap.get(x)) for x in
             (cv2.CAP_PROP_FRAME_WIDTH,
             cv2.CAP_PROP_FRAME_HEIGHT,
             cv2.CAP_PROP_FPS))
vw = cv2.VideoWriter(f"results_{s_mode}.avi",
                     cv2.VideoWriter_fourcc(*"mp4v"),
                     fps,
                     (w, h))
while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break

    results = model.track(im0,persist=True)[0]
    annotator = Annotator(im0)
    if results.boxes.id is not None:
        # Extract box, trackid, class and masks
        boxes = results.boxes.xyxy.tolist()
        tids = results.boxes.id.int().tolist()
        clss = results.boxes.id.int().tolist()
        masks = results.masks

        # preprocessing of mask before plotting
        img = LetterBox(masks.shape[1:])(
            image = annotator.result())
        img_gpu = (torch.as_tensor
                   (img, dtype=torch.float16,
                    device=masks.data.device)
                   .permute(2,0,1).flip(0)
                   .contiguous()/255)

        # Masks plotting
        annotator.masks(masks.data, colors=[
            colors(x, True)
            for x in (tids if s_mode==1 else clss)],
                        img_gpu=img_gpu)

        # Bounding box plotting
        for b,t,c in zip(boxes, tids, clss):
            annotator.box_label(
                b,
                color = colors(t if s_mode==1 else c , True),
                label=names[c])

        cv2.imshow("Image segmentation", im0)
        vw.write(im0)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
    cap.release()
    cv2.destroyAllWindows()

**Reply by sjzypf** [reply]
im_gpu你写成了img_gpu

**Comment by mdx123456** [comment]
您好，利用官方给的代码进行了语义分割的测试，报了错误：TypeError: masks() got an unexpected keyword argument 'img_gpu'
请给出一个解决方案指点一下，非常感谢！下面是我的完整测试代码：
import cv2
import torch
from ultralytics import YOLO
from ultralytics.utils.plotting import colors, Annotator
from ultralytics.data.augment import LetterBox

s_mode = 1  # 0: semantic, 1: instance
cap = cv2.VideoCapture('1.mp4')

model = YOLO('yolo11n-seg.pt')
names = model.names

w, h, fps = (int(cap.get(x)) for x in
             (cv2.CAP_PROP_FRAME_WIDTH,
             cv2.CAP_PROP_FRAME_HEIGHT,
             cv2.CAP_PROP_FPS))
vw = cv2.VideoWriter(f"results_{s_mode}.avi",
                     cv2.VideoWriter_fourcc(*"mp4v"),
                     fps,
                     (w, h))
while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break

    results = model.track(im0,persist=True)[0]
    annotator = Annotator(im0)
    if results.boxes.id is not None:
        # Extract box, trackid, class and masks
        boxes = results.boxes.xyxy.tolist()
        tids = results.boxes.id.int().tolist()
        clss = results.boxes.id.int().tolist()
        masks = results.masks

        # preprocessing of mask before plotting
        img = LetterBox(masks.shape[1:])(
            image = annotator.result())
        img_gpu = (torch.as_tensor
                   (img, dtype=torch.float16,
                    device=masks.data.device)
                   .permute(2,0,1).flip(0)
                   .contiguous()/255)

        # Masks plotting
        annotator.masks(masks.data, colors=[
            colors(x, True)
            for x in (tids if s_mode==1 else clss)],
                        img_gpu=img_gpu)

        # Bounding box plotting
        for b,t,c in zip(boxes, tids, clss):
            annotator.box_label(
                b,
                color = colors(t if s_mode==1 else c , True),
                label=names[c])

        cv2.imshow("Image segmentation", im0)
        vw.write(im0)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
    cap.release()
    cv2.destroyAllWindows()

联系邮箱：1059071410@qq.com

**Reply by glenn-jocher** [reply]
您好 @mdx123456，

看起来您遇到的错误是因为 `Annotator.masks()` 方法不接受 `img_gpu` 参数。这可能是因为您使用的 Ultralytics 版本与示例代码不匹配。

请尝试修改您的代码，移除 `img_gpu` 参数：

```python
# Masks plotting
annotator.masks(masks.data, colors=[
    colors(x, True)
    for x in (tids if s_mode==1 else clss)])
```

如果这仍然不起作用，您可以尝试更新到最新版本的 Ultralytics：

```
pip install -U ultralytics
```

此外，您也可以直接使用 YOLO 的内置可视化功能。例如：

```python
results = model.track(im0, persist=True)[0]
annotated_frame = results.plot()
cv2.imshow("Image segmentation", annotated_frame)
```

这种方法能够自动处理掩码和边界框的可视化，并且更加简洁。

希望这对您有所帮助！

**Comment by mdx123456** [comment]
你好，观看了官方测试的语义分割视频，想要利用官方源码实现语义分割的预测，采用pt或者onnx模型文件均可，按照工作人员建议更新ultralytics为最新版本也没解决，运行时报错TypeError: masks() got an unexpected keyword argument 'img_gpu' ，下面给出视频中的测试完整代码，请给出修改意见，非常感谢！
import cv2
import torch
from ultralytics import YOLO
from ultralytics.utils.plotting import colors, Annotator
from ultralytics.data.augment import LetterBox

s_mode = 1  # 0: semantic, 1: instance
cap = cv2.VideoCapture('1.mp4')

model = YOLO('yolo11n-seg.pt')
names = model.names

w, h, fps = (int(cap.get(x)) for x in
             (cv2.CAP_PROP_FRAME_WIDTH,
             cv2.CAP_PROP_FRAME_HEIGHT,
             cv2.CAP_PROP_FPS))
vw = cv2.VideoWriter(f"results_{s_mode}.avi",
                     cv2.VideoWriter_fourcc(*"mp4v"),
                     fps,
                     (w, h))
while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break

    results = model.track(im0,persist=True)[0]
    annotator = Annotator(im0)
    if results.boxes.id is not None:
        # Extract box, trackid, class and masks
        boxes = results.boxes.xyxy.tolist()
        tids = results.boxes.id.int().tolist()
        clss = results.boxes.id.int().tolist()
        masks = results.masks

        # preprocessing of mask before plotting
        img = LetterBox(masks.shape[1:])(
            image = annotator.result())
        img_gpu = (torch.as_tensor
                   (img, dtype=torch.float16,
                    device=masks.data.device)
                   .permute(2,0,1).flip(0)
                   .contiguous()/255)

        # Masks plotting
        annotator.masks(masks.data, colors=[
            colors(x, True)
            for x in (tids if s_mode==1 else clss)],
                        img_gpu=img_gpu)
        # results = model.track(im0, persist=True)[0]
        # annotated_frame = results.plot()
        # cv2.imshow("Image segmentation", annotated_frame)

        # Bounding box plotting
        for b,t,c in zip(boxes, tids, clss):
            annotator.box_label(
                b,
                color = colors(t if s_mode==1 else c , True),
                label=names[c])

        cv2.imshow("Image segmentation", im0)
        vw.write(im0)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
    cap.release()
    cv2.destroyAllWindows()
联系邮箱：1059071410@qq.com

**Reply by glenn-jocher** [reply]
您好 @mdx123456，

我看到您在尝试实现基于YOLO11的语义分割预测，并遇到了`img_gpu`参数错误。这个问题可能是由于示例代码与当前Ultralytics版本的API不匹配造成的。

我建议您尝试以下两种解决方案：

1. 使用更简化的代码方式，直接利用内置的可视化功能：

```python
import cv2
from ultralytics import YOLO

s_mode = 1  # 0: semantic, 1: instance
cap = cv2.VideoCapture('1.mp4')
model = YOLO('yolo11n-seg.pt')

w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))
vw = cv2.VideoWriter(f"results_{s_mode}.avi", cv2.VideoWriter_fourcc(*"mp4v"), fps, (w, h))

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break
        
    results = model.track(im0, persist=True)[0]
    annotated_frame = results.plot()  # 使用内置绘图功能
    
    cv2.imshow("Image segmentation", annotated_frame)
    vw.write(annotated_frame)
    
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
vw.release()
cv2.destroyAllWindows()
```

2. 如果您确实需要自定义可视化，可以检查当前Annotator类的masks方法签名：

```python
# 查看当前API的masks方法参数
from ultralytics.utils.plotting import Annotator
help(Annotator.masks)
```

然后根据实际API调整您的代码调用方式。

希望这对您有所帮助！如果您仍然遇到困难，请考虑查看最新的[Ultralytics文档](https://docs.ultralytics.com/tasks/segment/)获取更多信息。

**Comment by mdx123456** [comment]
您好，请问一下能否帮忙编写一个简单的Winform中实现C#调用官网yolo11语义分割和实例分割的onnx模型进行推理的完整代码？选择图片，点击按钮进行推理，展示推理后的效果图。
非常感谢!
联系邮箱：1059071410@qq.com

**Reply by glenn-jocher** [reply]
您好 @mdx123456，

虽然我无法在这里提供完整的C# Winform应用程序代码，但我可以指导您如何在C#中使用YOLO11 ONNX模型进行语义分割和实例分割。

您可以使用Microsoft.ML.OnnxRuntime库来加载和运行ONNX模型。以下是一个基本的实现思路：

1. 创建一个Winform应用，添加图片选择按钮、推理按钮和图片显示控件
2. 添加NuGet包：`Microsoft.ML.OnnxRuntime`和`OpenCvSharp4`（用于图像处理）
3. 使用OnnxRuntime加载模型并进行推理

建议您查看以下资源：
- [使用OnnxRuntime在.NET中运行ONNX模型](https://onnxruntime.ai/docs/get-started/with-csharp.html)
- Ultralytics的[ONNX导出和推理指南](https://docs.ultralytics.com/integrations/onnx/)

对于完整实现，您可能需要参考我们的Python示例并将处理逻辑转换为C#。如有更多Ultralytics特定问题，欢迎在GitHub讨论中继续交流。

**Comment by zhangjiageng** [comment]
您好，我分别用yolov5、v8、v11进行缺陷检测。我的数据集是大小为1900*210的二值图像，图像结构简单，图像主体是一个具有一定宽度的白色半圆，背景为纯黑色，人工在半圆上添加了一些形状、大小不等的黑色缺陷。现在检测结果是v5优于v11优于v8，请问这是为什么

**Comment by fn9170** [comment]
Wow

**Reply by sjzypf** [reply]
wow

**Reply by glenn-jocher** [reply]
您好 @zhangjiageng，

您观察到的性能差异很有趣，这可能与几个因素有关：

1. **数据集特性匹配**：您的数据集是1900×210的二值图像，结构相对简单。YOLOv5可能在这种特定的图像尺寸和简单结构上更适合，而较新的模型可能针对更复杂的自然图像场景进行了优化。

2. **模型复杂度**：YOLO11和YOLOv8引入了更复杂的架构改进，但在简单的二值缺陷检测任务中，这些改进可能并不能带来优势，反而可能导致过拟合。

3. **训练策略**：不同版本的默认训练超参数和数据增强策略可能对您的特定数据集产生不同影响。

建议您尝试：
- 调整图像尺寸（imgsz参数）以匹配您的数据集比例
- 针对您的数据集调整训练超参数
- 使用较小的模型变体（如yolo11n.pt）可能更适合简单任务

每个YOLO版本都有其优势，选择最适合您具体任务的版本才是最重要的。

**Reply by zhangjiageng** [reply]
谢谢您的回复！！！

**Reply by glenn-jocher** [reply]
您不客气！很高兴能帮助解决您的问题。如果您在进一步优化缺陷检测模型时遇到其他问题，欢迎随时在这里提问。祝您的项目顺利！

**Comment by zzw725** [comment]
WARNING ⚠️ D:\pytorch1\dy_test\deal_detect\ultralytics\ultralytics\weight\best.pt appears to require 'ultralytics.nn.modules.ShuffleNetV2', which is not in ultralytics requirements.
AutoInstall will run now for 'ultralytics.nn.modules.ShuffleNetV2' but this feature will be removed in the future.
Recommend fixes are to train a new model using the latest 'ultralytics' package or to run a command with an official YOLOv8 model, i.e. 'yolo predict model=yolov8n.pt'
WARNING ⚠️ D:\pytorch1\dy_test\deal_detect\ultralytics\ultralytics\weight\best.pt appears to require 'ultralytics.nn.modules.ShuffleNetV2', which is not in ultralytics requirements.
AutoInstall will run now for 'ultralytics.nn.modules.ShuffleNetV2' but this feature will be removed in the future.
Recommend fixes are to train a new model using the latest 'ultralytics' package or to run a command with an official YOLOv8 model, i.e. 'yolo predict model=yolov8n.pt'
Traceback (most recent call last):
  File "D:\pytorch1\dy_test\deal_detect\ultralytics\ultralytics\nn\tasks.py", line 673, in torch_safe_load
    ckpt = torch.load(file, map_location="cpu")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\anaconda\envs\yolo\Lib\site-packages\torch\serialization.py", line 1026, in load
    return _load(opened_zipfile,
           ^^^^^^^^^^^^^^^^^^^^^
  File "D:\anaconda\envs\yolo\Lib\site-packages\torch\serialization.py", line 1438, in _load
    result = unpickler.load()
             ^^^^^^^^^^^^^^^^
  File "D:\anaconda\envs\yolo\Lib\site-packages\torch\serialization.py", line 1431, in find_class
    return super().find_class(mod_name, name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'ultralytics.nn.modules.ShuffleNetV2'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\pytorch1\dy_test\deal_detect\ultralytics\ultralytics\detect_control.py", line 7, in <module>
    model = YOLO(weight_path)
            ^^^^^^^^^^^^^^^^^
  File "D:\pytorch1\dy_test\deal_detect\ultralytics\ultralytics\engine\model.py", line 134, in __init__
    self._load(model, task=task)
  File "D:\pytorch1\dy_test\deal_detect\ultralytics\ultralytics\engine\model.py", line 215, in _load
    self.model, self.ckpt = attempt_load_one_weight(weights)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\pytorch1\dy_test\deal_detect\ultralytics\ultralytics\nn\tasks.py", line 748, in attempt_load_one_weight
    ckpt, weight = torch_safe_load(weight)  # load ckpt
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\pytorch1\dy_test\deal_detect\ultralytics\ultralytics\nn\tasks.py", line 693, in torch_safe_load
    ckpt = torch.load(file, map_location="cpu")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\anaconda\envs\yolo\Lib\site-packages\torch\serialization.py", line 1026, in load
    return _load(opened_zipfile,
           ^^^^^^^^^^^^^^^^^^^^^
  File "D:\anaconda\envs\yolo\Lib\site-packages\torch\serialization.py", line 1438, in _load
    result = unpickler.load()
             ^^^^^^^^^^^^^^^^
  File "D:\anaconda\envs\yolo\Lib\site-packages\torch\serialization.py", line 1431, in find_class
    return super().find_class(mod_name, name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'ultralytics.nn.modules.ShuffleNetV2'

Everyone！
When I make improvements to the original module, the prediction will report the same error as above. When forecasting, the nn in the project is not called, but the nn in the virtual environment is called? Why is this happening? If you change the nn in the virtual environment, it will work normally, but how to avoid calling the nn in the virtual environment, but call the nn in the project.

**Reply by glenn-jocher** [reply]
您好 @zzw725，

这个问题是因为您修改了本地的ultralytics代码添加了ShuffleNetV2模块，但Python仍在使用pip安装的ultralytics包，而不是您的本地修改版本。

解决方案是以开发模式安装您的本地ultralytics版本：

```bash
# 首先卸载现有的ultralytics包
pip uninstall ultralytics

# 然后在您的项目目录中以可编辑模式安装
cd /path/to/your/ultralytics/project
pip install -e .
```

这样Python就会使用您本地修改的代码，包括您添加的ShuffleNetV2模块。

另外，建议您在训练新模型时使用最新的官方ultralytics包，这样可以避免版本兼容性问题。如果您需要自定义模块，可以考虑通过配置文件的方式添加，而不是直接修改源码。

希望这能解决您的问题！

**Comment by zizhennini** [comment]
您好，我们是在https://ultralytics.com/上传自己的数据集然后训练的，但为什么数据集里面没有chair也是别出来了，汉堡🍔识别成三明治且正确率在0.7到0.95

**Comment by daisheng123** [comment]
I looked at the parsing function in deepstream-yolo-pose, and it only handles single-class human pose estimation. How should it be parsed for multiple classes?

**Reply by glenn-jocher** [reply]
Ultralytics pose models assume a single keypoint layout (one `kpt_shape`/skeleton), so you can only do “multi-class pose” with one DeepStream parser if all classes share the same keypoint definition; in that case, parse keypoints exactly the same way and just use each detection’s predicted class id (`cls`) to tag/group the instance (see the [Ultralytics pose task docs](https://docs.ultralytics.com/tasks/pose/)). If your classes need different keypoint counts/semantics, you’ll need separate pose models/engines (or run detection first, then a class-specific pose model), since one output tensor shape can’t represent multiple skeletons cleanly.
